File: backend/src/ws_contracts.py
"""
ws_contracts.py — Pydantic schemas for all outgoing WebSocket messages.

These schemas serve as the single source of truth for the WS message contract.
All broadcast/send calls in connection_manager and game_master should produce
payloads that conform to these types.

The frontend Zod schemas in frontend/src/types/wsSchemas.ts must mirror this file.
"""
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


# ── Narrative ────────────────────────────────────────────────────────────────

class NarrativeTokenMessage(BaseModel):
    type: Literal["narrative_token"]
    token: str


class StreamEndMessage(BaseModel):
    type: Literal["stream_end"]


class GmThinkingMessage(BaseModel):
    type: Literal["gm_thinking"]
    message: str


# ── Game State ───────────────────────────────────────────────────────────────

class GameEventMessage(BaseModel):
    type: Literal["game_event"]
    event: str
    payload: dict[str, Any]


class StateUpdateMessage(BaseModel):
    type: Literal["state_update"]
    delta: dict[str, Any]


class FullStateSyncMessage(BaseModel):
    type: Literal["full_state_sync"]
    state: dict[str, Any]
    world_state: dict[str, Any] | None = None


class HistorySyncMessage(BaseModel):
    type: Literal["history_sync"]
    entries: list[dict[str, Any]]


# ── Dice ─────────────────────────────────────────────────────────────────────

class DiceRollMessage(BaseModel):
    type: Literal["dice_roll"]
    player: str
    die: int = Field(gt=0)
    purpose: str
    result: int
    is_critical: bool = False
    is_fumble: bool = False


class RequestDiceRollMessage(BaseModel):
    type: Literal["request_dice_roll"]
    roll_id: str
    roll_type: str
    dc: int
    description: str


class DiceRollResolvedMessage(BaseModel):
    type: Literal["dice_roll_resolved"]
    roll_id: str
    verdict: Literal["accept_with_bonus", "accept_no_bonus", "reject", "reroll_requested"]
    circumstance_bonus: int
    final_result: int | None
    explanation: str


# ── Audio / Media ────────────────────────────────────────────────────────────

class AudioCueMessage(BaseModel):
    type: Literal["audio_cue"]
    cue: str


class BossMusicMessage(BaseModel):
    type: Literal["boss_music"]
    url: str | None = None
    tension_level: int | None = None
    intensity: Literal["high", "medium"] | None = None


class ImageReadyMessage(BaseModel):
    type: Literal["image_ready"]
    url: str
    subject: str


# ── Auth ─────────────────────────────────────────────────────────────────────

class TokenRefreshMessage(BaseModel):
    type: Literal["token_refresh"]
    access_token: str


# ── Isekai ───────────────────────────────────────────────────────────────────

class IsekaiConvocationMessage(BaseModel):
    type: Literal["isekai_convocation"]
    faction: str
    narrative: str
    secret_objective: str


class FactionObjectiveUpdateMessage(BaseModel):
    type: Literal["faction_objective_update"] = "faction_objective_update"
    faction: str
    objective: str
    status: Literal["in_progress", "completed", "failed"]
    cred_change: float


class DiceResultMessage(BaseModel):
    type: Literal["dice_result"] = "dice_result"
    player_id: str
    die: int = Field(gt=0)
    result: int


# ── Errors ───────────────────────────────────────────────────────────────────

class ErrorMessage(BaseModel):
    type: Literal["error"]
    message: str


class MilestoneMessage(BaseModel):
    type: Literal["milestone"] = "milestone"
    player_id: str
    milestones: list[str]


# ── Church Seals ─────────────────────────────────────────────────────────────

class SealEventMessage(BaseModel):
    type: Literal["seal_event"] = "seal_event"
    player_id: str
    action: Literal["granted", "revoked"]
    seal_type: str | None


# ── Union type for all outgoing messages ─────────────────────────────────────

OutgoingWSMessage = (
    NarrativeTokenMessage
    | StreamEndMessage
    | GmThinkingMessage
    | GameEventMessage
    | StateUpdateMessage
    | FullStateSyncMessage
    | HistorySyncMessage
    | DiceRollMessage
    | RequestDiceRollMessage
    | DiceRollResolvedMessage
    | AudioCueMessage
    | BossMusicMessage
    | ImageReadyMessage
    | TokenRefreshMessage
    | IsekaiConvocationMessage
    | FactionObjectiveUpdateMessage
    | DiceResultMessage
    | MilestoneMessage
    | SealEventMessage
    | ErrorMessage
)

File: frontend/src/types/wsContracts.ts
/**
 * wsContracts.ts — Zod schemas for all incoming WebSocket messages.
 *
 * Mirror of backend/src/ws_contracts.py.
 * When adding a new WS event type, update BOTH files simultaneously.
 *
 * Usage:
 *   import { parseWSMessage } from './wsContracts'
 *   const parsed = parseWSMessage(rawJson)
 *   if (parsed.type === 'narrative_token') { ... }
 */
import { z } from 'zod'

// ── Narrative ────────────────────────────────────────────────────────────────

export const NarrativeTokenSchema = z.object({
  type: z.literal('narrative_token'),
  token: z.string(),
})

export const StreamEndSchema = z.object({
  type: z.literal('stream_end'),
})

export const GmThinkingSchema = z.object({
  type: z.literal('gm_thinking'),
  message: z.string(),
})

// ── Game State ───────────────────────────────────────────────────────────────

export const GameEventSchema = z.object({
  type: z.literal('game_event'),
  event: z.string(),
  payload: z.record(z.string(), z.unknown()),
})

export const StateUpdateSchema = z.object({
  type: z.literal('state_update'),
  delta: z.record(z.string(), z.unknown()),
})

export const FullStateSyncSchema = z.object({
  type: z.literal('full_state_sync'),
  state: z.record(z.string(), z.unknown()),
  world_state: z.record(z.string(), z.unknown()).optional(),
})

export const HistorySyncSchema = z.object({
  type: z.literal('history_sync'),
  entries: z.array(z.object({
    role: z.enum(['user', 'assistant']),
    content: z.string(),
    turn_number: z.number().int().optional(),
  })),
})

// ── Dice ─────────────────────────────────────────────────────────────────────

export const DiceRollSchema = z.object({
  type: z.literal('dice_roll'),
  player: z.string(),
  die: z.number().int().positive(),
  purpose: z.string(),
  result: z.number().int(),
  is_critical: z.boolean().default(false),
  is_fumble: z.boolean().default(false),
})

export const RequestDiceRollSchema = z.object({
  type: z.literal('request_dice_roll'),
  roll_id: z.string(),
  roll_type: z.string(),
  dc: z.number().int(),
  description: z.string(),
})

export const DiceRollResolvedSchema = z.object({
  type: z.literal('dice_roll_resolved'),
  roll_id: z.string(),
  verdict: z.enum(['accept_with_bonus', 'accept_no_bonus', 'reject', 'reroll_requested']),
  circumstance_bonus: z.number().int(),
  final_result: z.number().int().nullable(),
  explanation: z.string(),
})

// ── Audio / Media ────────────────────────────────────────────────────────────

export const AudioCueSchema = z.object({
  type: z.literal('audio_cue'),
  cue: z.string(),
})

export const BossMusicSchema = z.object({
  type: z.literal('boss_music'),
  tension_level: z.number().int().optional(),
  intensity: z.enum(['high', 'medium']).optional(),
  url: z.string().optional(),
})

export const ImageReadySchema = z.object({
  type: z.literal('image_ready'),
  url: z.string().url(),
  subject: z.string(),
})

// ── Auth ─────────────────────────────────────────────────────────────────────

export const TokenRefreshSchema = z.object({
  type: z.literal('token_refresh'),
  access_token: z.string(),
})

// ── Isekai ───────────────────────────────────────────────────────────────────

export const IsekaiConvocationSchema = z.object({
  type: z.literal('isekai_convocation'),
  faction: z.string(),
  narrative: z.string(),
  secret_objective: z.string(),
})

export const FactionObjectiveUpdateSchema = z.object({
  type: z.literal('faction_objective_update'),
  faction: z.string(),
  objective: z.string(),
  status: z.enum(['in_progress', 'completed', 'failed']),
  cred_change: z.number(),
})

// ── Errors ───────────────────────────────────────────────────────────────────

export const ErrorMessageSchema = z.object({
  type: z.literal('error'),
  message: z.string(),
})

// ── Discriminated union of all incoming messages ──────────────────────────────

export const WSMessageSchema = z.discriminatedUnion('type', [
  NarrativeTokenSchema,
  StreamEndSchema,
  GmThinkingSchema,
  GameEventSchema,
  StateUpdateSchema,
  FullStateSyncSchema,
  HistorySyncSchema,
  DiceRollSchema,
  RequestDiceRollSchema,
  DiceRollResolvedSchema,
  AudioCueSchema,
  BossMusicSchema,
  ImageReadySchema,
  TokenRefreshSchema,
  IsekaiConvocationSchema,
  FactionObjectiveUpdateSchema,
  ErrorMessageSchema,
])

export type WSMessage = z.infer<typeof WSMessageSchema>

/**
 * Parse and validate a raw WebSocket message payload.
 * Returns the typed message or throws a ZodError on invalid shape.
 */
export function parseWSMessage(raw: unknown): WSMessage {
  return WSMessageSchema.parse(raw)
}

/**
 * Safe version — returns null on parse failure instead of throwing.
 * Use for non-critical paths where a malformed message should be silently ignored.
 */
export function safeParseWSMessage(raw: unknown): WSMessage | null {
  const result = WSMessageSchema.safeParse(raw)
  return result.success ? result.data : null
}
